// nxpm/lib/commands/sandbox/pull.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const sandbox_pull_1 = require("../../lib/sandbox/sandbox-pull");
const utils_1 = require("../../utils");
class SandboxPull extends utils_1.BaseCommand {
    async run() {
        const { flags } = this.parse(SandboxPull);
        await sandbox_pull_1.sandboxPull({
            force: flags.force,
            refresh: flags.refresh,
            remove: flags.remove,
            config: this.config,
            userConfig: this.userConfig,
        });
    }
}
exports.default = SandboxPull;
SandboxPull.description = 'Pull images of sandboxes';
SandboxPull.flags = {
    force: command_1.flags.boolean({
        char: 'f',
        description: 'Force removal of the sandboxes',
        default: false,
    }),
    help: command_1.flags.help({ char: 'h' }),
    refresh: command_1.flags.boolean({
        char: 'r',
        description: 'Refresh the list of sandboxes',
        default: false,
    }),
    remove: command_1.flags.boolean({
        char: 'm',
        description: 'Remove all of the sandboxes before pulling',
        default: false,
    }),
};
