// nxpm/lib/commands/projects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const projects_1 = require("../lib/projects/projects");
const utils_1 = require("../utils");
class Projects extends utils_1.BaseCommand {
    async run() {
        const { args, flags } = this.parse(Projects);
        await projects_1.projects({ cwd: flags.cwd, dryRun: false, userConfig: this.userConfig, config: this.config }, args.projectName, args.target);
    }
}
exports.default = Projects;
Projects.aliases = ['p'];
Projects.description = 'Interactive menu to run builders and schematics for projects';
Projects.flags = {
    cwd: command_1.flags.string({
        char: 'c',
        description: 'Current working directory',
        default: process.cwd(),
    }),
    help: command_1.flags.help({ char: 'h' }),
};
Projects.args = [
    {
        name: 'projectName',
        description: 'The name of the project you want to operate on',
        required: false,
    },
    {
        name: 'target',
        description: 'The target to run (build, serve, test, etc)',
        required: false,
    },
];
